### PR TITLE
GLOB-5138: Add a URIField

### PR DESCRIPTION
**Pull request by jessemyers**
Includes:
 -  Reorganization of custom fields
 -  Addition of the `rfc3986` library and a URIField
 -  Elimination of the `urltools` library that makes an HTTP request to the Internet on load
 -  Movement of URIMatcher to an importable location

### CODE REVIEW

File: microcosm_flask/fields.py [hunk 1]
```diff
@@ -1,128 +0,0 @@
-"""
-Custom fields.
```
Comment (jessemyers) [inline microcosm_flask/fields.py]: Moved to a sub-package.

File: microcosm_flask/fields/uri_field.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+"""
+A URI valued (and normalized) field.
```
Comment (jessemyers) [inline microcosm_flask/fields/uri_field.py]: This is new.

File: microcosm_flask/fields/uri_field.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+
+def normalize_uri_result(uri):
+    """
+    Normalize a URI (And return a URIResult)
+
+    """
+    ref = uri_reference(uri).normalize()
+
+    return ref._replace(
```
Comment (jessemyers) [inline microcosm_flask/fields/uri_field.py]: I'm still not super happy with the `rfc3986` library; it doesn't quite normalize as I'd hope so I wrote some extra processing below.

File: microcosm_flask/matchers.py [hunk 4]
```diff
@@ -63,3 +67,30 @@ def schema_class(self):
 
     def describe_to(self, description):
         description.append_text("expected {}".format(prettify(self.expected)))
+
+
+class URIMatcher(BaseMatcher):
```
Comment (jessemyers) [inline microcosm_flask/matchers.py]: The previous `URLMatcher` (name changed  here) was in the `test` tree so it couldn't be imported in other projects. (We may not need to use it BTW if we normalize URIs on the input side via the new field)

File: microcosm_flask/tests/fields/test_uri_field.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+from marshmallow import Schema, ValidationError
+from microcosm_flask.fields import URIField
+from microcosm_flask.fields.uri_field import normalize_uri
+
+
+class URISchema(Schema):
+    foo = URIField(include_query=True)
+
+
+def test_normalize_uri():
```
Comment (jessemyers) [inline microcosm_flask/tests/fields/test_uri_field.py]: Test cases...

**Comment by netama** [comment]
LGTM. Would adding another field type solve the schema validation error in https://dodona.dev.globality.io:5422/api/v1/swagger (for using a schema in the query)?

**Comment by jessemyers** [comment]
@netama I'm not sure about that schema validation error.

**Pull request closed without merging**